### PR TITLE
Skip sequence parallel operations during eval

### PR DESCRIPTION
**Pull request by jp1924**
hi deepspeed!

[Make it possible to evaluate when using sequence parallel in HF Trainer #43517](https://github.com/huggingface/transformers/pull/43517)

I initially opened a PR in transformers to prevent errors that occur when running eval with deepspeed sequence parallel applied, but there was feedback that this should be handled in deepspeed rather than in transformers or accelerate, so I'm opening this PR here.

If you have any questions, feel free to ask.

### CODE REVIEW

**Comment by jp1924** [comment]
CC: @kashif @SunMarc

**Comment by tohtana** [comment]
Hi @jp1924,
Thanks for opening this PR! I agree the use case is important. That said, I’d like to be cautious about forcing this behavior by default. I opened another PR that lets users opt in/out of SP in eval mode.
Do you think that would work for your use case?

**Comment by jp1924** [comment]
@tohtana Oh! Right. This should be fine. And looking at the code, the test seems fine too.
But this is a personal request—could I submit a PR separately and try to become a contributor?

**Comment by tohtana** [comment]
Okay, you can merge my [PR branch](https://github.com/tohtana/DeepSpeed/tree/tohtana/allow_disable_sp_in_eval_mode). But can you try it on your end? It would be very helpful to verify if the approach works in your actual case.

**Comment by jp1924** [comment]
@tohtana Ah, I've checked it. I'll test it and share the results with you.

**Comment by jp1924** [comment]
@tohtana, it works normally.

However, to use it normally, the following changes are needed.

1. For multi-modal configs divided into vision_config and text_config, it does not work normally.
   - Added hardcode `hf_model_config = getattr(hf_model_config, "text_config", hf_model_config)` to the register_with_transformers method to make it work.
   - But this seems to be a change that should be added to accelerate, not DeepSpeed.

2. Enable disable_in_eval in accelerate
   - Hardcoded disable_in_eval to True in `UlyssesSPAttentionHF.register_with_transformers` in accelerate > accelerator.py > _prepare_deepspeed.
   - But this needs to be discussed with the accelerate team on how to add disable_in_eval.

3. Add self.model.training as in this PR https://github.com/huggingface/transformers/pull/43517
   - In transformers > trainer.py > _inner_training_loop, it should decide whether to take `self._deepspeed_sp_compute_loss`.

When these 3 changes are applied, it works normally. 
For now, it seems that everything that can be handled on the DeepSpeed side is done with this?

**Comment by jp1924** [comment]
## SP Enabled
```
***** Running training *****
  Num examples = 92
  Num Epochs = 2
  Instantaneous batch size per device = 1
  Total train batch size (w. parallel, distributed & accumulation) = 1
  Gradient Accumulation steps = 1
  Total optimization steps = 184
  Number of trainable parameters = 12,187,325,040
{'loss': '19.36', 'grad_norm': '697', 'learning_rate': '0', 'num_tokens': '1.636e+04', 'mean_token_accuracy': '0.02733', 'epoch': '0.01087', 'num_input_tokens_seen': 16362, 'train_runtime': '84.78', 'train_tokens_per_second': '193'}
  1%|▌                                                                                                                | 1/184 [01:24<4:18:28, 84.75s/it]
***** Running Evaluation *****
Num examples = 586
Batch size = 12
{'eval_loss': '1.362', 'eval_runtime': '51.7', 'eval_samples_per_second': '11.34', 'eval_steps_per_second': '0.484', 'eval_num_tokens': '1.636e+04', 'epoch': '0.01087', 'num_input_tokens_seen': 16362}                                                                                                        
{'loss': '21.03', 'grad_norm': '1069', 'learning_rate': '3.333e-06', 'num_tokens': '3.256e+04', 'mean_token_accuracy': '0.01765', 'epoch': '0.02174', 'num_input_tokens_seen': 32562, 'train_runtime': '198.7', 'train_tokens_per_second': '163.9'}                                                             
  1%|█▏                                                                                                              | 2/184 [03:18<5:09:07, 101.91s/it]
```

## SP Disabled
```
***** Running training *****
  Num examples = 2,474
  Num Epochs = 2
  Instantaneous batch size per device = 32
  Total train batch size (w. parallel, distributed & accumulation) = 64
  Gradient Accumulation steps = 1
  Total optimization steps = 78
  Number of trainable parameters = 12,187,325,040
{'loss': '1.317', 'grad_norm': '45.93', 'learning_rate': '0', 'num_tokens': '3.107e+04', 'mean_token_accuracy': '0.7753', 'epoch': '0.02564', 'num_input_tokens_seen': 31066, 'train_runtime': '83.31', 'train_tokens_per_second': '372.9'}
  1%|█▍                                                                                                                | 1/78 [01:23<1:46:52, 83.28s/it]
***** Running Evaluation *****
Num examples = 586
Batch size = 12
{'eval_loss': '1.338', 'eval_runtime': '49.48', 'eval_samples_per_second': '11.84', 'eval_steps_per_second': '0.505', 'eval_num_tokens': '3.107e+04', 'epoch': '0.02564', 'num_input_tokens_seen': 31066}                                                                                                       
{'loss': '1.277', 'grad_norm': '45.97', 'learning_rate': '6.667e-06', 'num_tokens': '6.256e+04', 'mean_token_accuracy': '0.779', 'epoch': '0.05128', 'num_input_tokens_seen': 62555, 'train_runtime': '194.4', 'train_tokens_per_second': '321.8'}                                                              
  3%|██▉                                                                                                               | 2/78 [03:14<2:06:13, 99.65s/it]
```

**Comment by tohtana** [comment]
With these results, you mean the results are inconsistent?
You will need to change the inputs to the model depending on whether SP is enabled or not (SP expects inputs are partitioned).
I'm not very sure what change we need in accelerate side, but we might need to adjust the interface based on the requirements. If you are going to discuss there, let's leave this PR for a while.

**Comment by jp1924** [comment]
@tohtana 

Um, first, my explanation was insufficient, and please ignore the results above. I think I ran them incorrectly.
What I wanted to show through those results was that eval works without issues even with SP enabled, as demonstrated by the output logs.

I agree with your opinion. I think we need to ask the accelerate team as well. 
I'll CC you when I open a separate PR in accelerate.

<details>
<summary>What did I run incorrectly?</summary>

The experiment above was conducted with the code at https://github.com/jp1924/LLM/tree/main/src/sft. I added several features to make experimentation easier, but I should have removed them before running. Specifically, there's code that increases `max_length` by a multiple of `batch_size` and then patches `batch_size` to 1 when `seq_packing=True` and SP is enabled in sft_trainer. I should have disabled this, but I left it enabled.

</details>

So I ran the tests again and created A/B code to train `Gemma3-1B-it` on `trl-lib/alpaca-cleaned` with SP. 
I'm sharing the code and dependencies below for reproducibility.

<details>
<summary>Execution script</summary>

```python
import os

import torch
from datasets import load_dataset
from trl import SFTConfig, SFTTrainer
from trl.trainer.sft_trainer import DataCollatorForLanguageModeling

from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers import logging as hf_logging

logger = hf_logging.get_logger("transformers")
hf_logging.set_verbosity("INFO")
hf_logging.enable_default_handler()
hf_logging.enable_explicit_format()
os.environ["TOKENIZERS_PARALLELISM"] = "false"
os.environ["TORCHDYNAMO_VERBOSE"] = "1"
os.environ["TORCHDYNAMO_DISABLE"] = "1"


def main():
    class Collator(DataCollatorForLanguageModeling):
        def torch_call(self, examples):
            batch = super().torch_call(examples)
            input_ids, labels = batch["input_ids"], batch["labels"]
            attention_mask = batch["attention_mask"]
            sp_enabled = (
                trainer.accelerator.parallelism_config.sp_enabled
                if trainer.accelerator.parallelism_config
                else False
            )
            if sp_enabled and input_ids.shape[1] % trainer.args.world_size != 0:
                pad_length = trainer.args.world_size - (
                    input_ids.shape[1] % trainer.args.world_size
                )
                input_ids = torch.nn.functional.pad(
                    input_ids, (0, pad_length), value=self.pad_token_id
                )
                labels = torch.nn.functional.pad(labels, (0, pad_length), value=-100)
                attention_mask = torch.nn.functional.pad(
                    attention_mask, (0, pad_length)
                )

            batch["input_ids"] = input_ids
            batch["labels"] = labels
            batch["attention_mask"] = attention_mask
            return batch

    args = SFTConfig(
        output_dir="./please_rm",
        per_device_train_batch_size=8,
        per_device_eval_batch_size=4,
        optim="adamw_torch_fused",
        learning_rate=2e-5,
        num_train_epochs=1,
        logging_steps=1,
        do_eval=True,
        do_train=True,
        eval_strategy="steps",
        eval_steps=1,
        save_strategy="no",
        bf16=True,
        tf32=True,
        report_to="none",
        seed=42,
        remove_unused_columns=False,
    )

    dataset = load_dataset("trl-lib/alpaca-cleaned", split="train")
    train_dataset, valid_dataset = dataset.train_test_split(test_size=0.01).values()

    model_id = "google/gemma-3-1b-it"
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    model = AutoModelForCausalLM.from_pretrained(
        model_id, attn_implementation="flash_attention_2"
    )
    model = torch.compile(model)

    collator = Collator(pad_token_id=tokenizer.pad_token_id)

    trainer = SFTTrainer(
        args=args,
        model=model,
        train_dataset=train_dataset,
        eval_dataset=valid_dataset,
        processing_class=tokenizer,
        data_collator=collator,
    )
    trainer.train()


if __name__ == "__main__":
    main()
```

</details>

<details>
<summary>Accelerate config file (uncomment to enable SP)</summary>

```yaml
compute_environment: LOCAL_MACHINE
debug: false
downcast_bf16: 'no'
enable_cpu_affinity: false
distributed_type: DEEPSPEED
deepspeed_config:
  zero_stage: 3
  offload_param_device: cpu
  offload_optimizer_device: cpu
num_processes: 2
mixed_precision: bf16
# parallelism_config:
#   parallelism_config_sp_size: 2
#   parallelism_config_sp_backend: deepspeed
#   parallelism_config_sp_seq_length_is_variable: true
#   parallelism_config_sp_attn_implementation: flash_attention_2
```

</details>

<details>
<summary>Dependencies</summary>

```shell
pip install git+https://github.com/jp1924/transformers.git@6991e87f5f998dfed9a32ff658f8a3ec9d74b7d6
pip install git+https://github.com/jp1924/DeepSpeed.git@7b181c62d9b52ce01911ca0d2bed2d3d22ffeb30
pip install git+https://github.com/huggingface/accelerate.git@38dadd9537d43cb43e3149964c06a3f011b40aa6d
pip install git+https://github.com/huggingface/trl.git@ad91c6ffa91073684c4cf6dc2008e2994dd940e7
pip install torch==2.9.1
pip install flash-attn==2.8.3
```

</details>

**Run command:**
```shell
accelerate launch --config-file=<yaml-config-file> <python-code-file>
```

You need to hardcode `disable_in_eval=True` in DeepSpeed's SP config for this to work without errors.
_(To use sp normally in accelerate, hardcoding is required, so a pull request is needed for accelerate.)_

### Results (tested on 2x A100 80GB)

| Config | Eval Loss | Eval Mean Token Accuracy |
|--------|-----------|--------------------------|
| SP Enabled | 2.114 | 0.5973 |
| SP Disabled | 1.947 | 0.6009 |

<details>
<summary>full logs</summary>

### Sequence Parallel Enable True
```txt
***** Running training *****
  Num examples = 51,242
  Num Epochs = 1
  Instantaneous batch size per device = 8
  Total train batch size (w. parallel, distributed & accumulation) = 8
  Gradient Accumulation steps = 1
  Total optimization steps = 6,406
  Number of trainable parameters = 999,885,952
{'loss': '4.812', 'grad_norm': '99.95', 'learning_rate': '2e-05', 'entropy': '1.953', 'num_tokens': '1368', 'mean_token_accuracy': '0.4132', 'epoch': '0.0001561'}                                                                                                                                                  
  0%|                                                                                                                                                                                                                                                                           | 1/6406 [00:06<11:18:02,  6.35s/it]

***** Running Evaluation *****
  Num examples = 518
  Batch size = 4
{'eval_loss': '2.114', 'eval_runtime': '10.53', 'eval_samples_per_second': '49.2', 'eval_steps_per_second': '6.174', 'eval_entropy': '1.297', 'eval_num_tokens': '1368', 'eval_mean_token_accuracy': '0.5973', 'epoch': '0.0001561'}                                                                                
  0%|                                                                                                                                                                                                                                                                           | 1/6406 [00:16<11:18:02,  6.35s/it^CW0130
```
### Sequence Parallel Disable
```txt
***** Running training *****
  Num examples = 51,242
  Num Epochs = 1
  Instantaneous batch size per device = 8
  Total train batch size (w. parallel, distributed & accumulation) = 16
  Gradient Accumulation steps = 1
  Total optimization steps = 3,203
  Number of trainable parameters = 999,885,952
{'loss': '4.295', 'grad_norm': '105.5', 'learning_rate': '2e-05', 'entropy': '0.9453', 'num_tokens': '2253', 'mean_token_accuracy': '0.4877', 'epoch': '0.0003122'}                                                                                                                                                 
  0%|                                                                                                                                                                                                                                                                            | 1/3203 [00:05<4:55:07,  5.53s/it]

***** Running Evaluation *****
  Num examples = 518
  Batch size = 4
{'eval_loss': '1.947', 'eval_runtime': '11.77', 'eval_samples_per_second': '44', 'eval_steps_per_second': '5.521', 'eval_entropy': '1.284', 'eval_num_tokens': '2253', 'eval_mean_token_accuracy': '0.6009', 'epoch': '0.0003122'}                                                                                  
  0%|                                                                                                                                                                                                                                                                            | 1/3203 [00:17<4:55:07,  5.53s/it
```

</details>

The eval loss is nearly identical between SP enabled and disabled, which demonstrates that eval works correctly with SP. 
The training metrics differ as expected depending on SP status, but that's the expected behavior.

Therefore, there is no problem with the implementation.

**Comment by tohtana** [comment]
Thank you for the validation! The API design seems to work with accelerate as well. Let me merge this PR.

**Comment by jp1924** [comment]
thank you!
